### PR TITLE
Add named state support to the s3 backend

### LINKED ISSUE #12768: Cannot use envs with s3 backend for state
### Terraform Version
Terraform v0.9.0

### Affected Resource(s)
- Remote Backends
- State Environments

### Terraform Configuration Files
https://gist.github.com/sysadmiral/eb8b75124c44090d76884875f7431913

### Debug Output
https://gist.github.com/sysadmiral/39a2d06f1a59e6f46bb05645efc39530

### Expected Behavior
State Environments is usable with Remote Backends

### Actual Behavior
Once remote state is enable `terraform env` subcommands return `named states not supported` or similar.

### Steps to Reproduce
* Copy the config files
* Change profile in backend_config to a valid named profile
* Change client in terraform.tfvars to a valid named profile (${var.client} is used to set the profile in the aws provider)
* `terraform init -backend-config=backend_config` to setup remote state
* `terraform env list` to list environments

(since terraform init can be run at any time you can try the other way and setup envs first and do the init - the envs are created, remote state is initialized but you can't switch env or list envs)

### PR DESCRIPTION
**Pull request by jbardin**
This adds named state (environment) support to the S3 backend.

A state NAME will prepend the configured s3 key with `env:/NAME/`.
The default state will remain rooted in the bucket for backwards
compatibility.

Locks in DynamoDB use the S3 key as the as the primary key value, so
locking will work as expected for multiple states.

Fixes #12768

### CODE REVIEW

**Comment by jbardin** [comment]
```
$ TF_ACC=1 go test -v
=== RUN   TestBackend_impl
--- PASS: TestBackend_impl (0.00s)
=== RUN   TestBackendConfig
2017/03/22 16:47:03 [INFO] Ignoring AWS metadata API endpoint at default location as it doesn't return any instance-id
--- PASS: TestBackendConfig (0.70s)
=== RUN   TestBackend
2017/03/22 16:47:04 [INFO] Ignoring AWS metadata API endpoint at default location as it doesn't return any instance-id
2017/03/22 16:47:05 [DEBUG] Uploading remote state to S3: {
  Body: buffer(0xc4203ca000),
  Bucket: "terraform-remote-s3-test-58d2e2c7",
  ContentLength: 0,
  ContentType: "application/json",
  Key: "env:/foo/testState",
  ServerSideEncryption: "AES256"
}
2017/03/22 16:47:06 [DEBUG] Uploading remote state to S3: {
  Body: buffer(0xc4203cabd0),
  Bucket: "terraform-remote-s3-test-58d2e2c7",
  ContentLength: 0,
  ContentType: "application/json",
  Key: "env:/bar/testState",
  ServerSideEncryption: "AES256"
}
2017/03/22 16:47:07 [DEBUG] New state was assigned lineage "2c01796e-264e-47bb-9e30-6e5bc9a2eb1c"
2017/03/22 16:47:07 [TRACE] Preserving existing state lineage "bar"
2017/03/22 16:47:07 [DEBUG] Uploading remote state to S3: {
  Body: buffer(0xc420318660),
  Bucket: "terraform-remote-s3-test-58d2e2c7",
  ContentLength: 249,
  ContentType: "application/json",
  Key: "env:/bar/testState",
  ServerSideEncryption: "AES256"
}
--- PASS: TestBackend (8.23s)
	backend_test.go:122: creating S3 bucket terraform-remote-s3-test-58d2e2c7 in us-west-2
=== RUN   TestBackendLocked
2017/03/22 16:47:12 [INFO] Ignoring AWS metadata API endpoint at default location as it doesn't return any instance-id
2017/03/22 16:47:13 [INFO] Ignoring AWS metadata API endpoint at default location as it doesn't return any instance-id
2017/03/22 16:47:26 [DEBUG] Uploading remote state to S3: {
  Body: buffer(0xc420318d80),
  Bucket: "terraform-remote-s3-test-58d2e2cf",
  ContentLength: 0,
  ContentType: "application/json",
  Key: "env:/foo/testState",
  ServerSideEncryption: "AES256"
}
2017/03/22 16:47:27 [DEBUG] Uploading remote state to S3: {
  Body: buffer(0xc42037c090),
  Bucket: "terraform-remote-s3-test-58d2e2cf",
  ContentLength: 0,
  ContentType: "application/json",
  Key: "env:/bar/testState",
  ServerSideEncryption: "AES256"
}
2017/03/22 16:47:27 [DEBUG] New state was assigned lineage "b2a83f1f-b037-4fd9-8dec-5187d3f99446"
2017/03/22 16:47:27 [TRACE] Preserving existing state lineage "bar"
2017/03/22 16:47:27 [DEBUG] Uploading remote state to S3: {
  Body: buffer(0xc42037dec0),
  Bucket: "terraform-remote-s3-test-58d2e2cf",
  ContentLength: 249,
  ContentType: "application/json",
  Key: "env:/bar/testState",
  ServerSideEncryption: "AES256"
}
--- PASS: TestBackendLocked (24.99s)
	backend_test.go:122: creating S3 bucket terraform-remote-s3-test-58d2e2cf in us-west-2
	testing.go:170: TestBackend: testing state locking for *s3.Backend
=== RUN   TestRemoteClient_impl
--- PASS: TestRemoteClient_impl (0.00s)
=== RUN   TestRemoteClient
2017/03/22 16:47:37 [INFO] Ignoring AWS metadata API endpoint at default location as it doesn't return any instance-id
2017/03/22 16:47:38 [DEBUG] New state was assigned lineage "972d3bd7-9bd0-4bc5-bae4-1f4832de7e49"
2017/03/22 16:47:38 [TRACE] Preserving existing state lineage "972d3bd7-9bd0-4bc5-bae4-1f4832de7e49"
2017/03/22 16:47:38 [DEBUG] Uploading remote state to S3: {
  Body: buffer(0xc42046e9c0),
  Bucket: "terraform-remote-s3-test-58d2e2e8",
  ContentLength: 642,
  ContentType: "application/json",
  Key: "testState",
  ServerSideEncryption: "AES256"
}
--- PASS: TestRemoteClient (4.41s)
	backend_test.go:122: creating S3 bucket terraform-remote-s3-test-58d2e2e8 in us-west-2
=== RUN   TestRemoteClientLocks
2017/03/22 16:47:41 [INFO] Ignoring AWS metadata API endpoint at default location as it doesn't return any instance-id
2017/03/22 16:47:42 [INFO] Ignoring AWS metadata API endpoint at default location as it doesn't return any instance-id
--- PASS: TestRemoteClientLocks (19.14s)
	backend_test.go:122: creating S3 bucket terraform-remote-s3-test-58d2e2ed in us-west-2
PASS
ok  	github.com/hashicorp/terraform/backend/remote-state/s3	57.490s

File: backend/remote-state/s3/backend_state.go [hunk 1]
```diff
@@ -1,27 +1,112 @@
...
 	"github.com/hashicorp/terraform/backend"
 	"github.com/hashicorp/terraform/state"
 	"github.com/hashicorp/terraform/state/remote"
 )
 
 const (
-	keyEnvPrefix = "-env:"
+	// This will be used as directory name, the odd looking colon is simply to
+	// reduce the chance of name conflicts with existing objects.
+	keyEnvPrefix = "env:"
```
Comment (mitchellh) [inline backend/remote-state/s3/backend_state.go]: Is the removal of "-" significant at all?

Comment (jbardin) [inline backend/remote-state/s3/backend_state.go]: Nope. The `-` was just in a copy&paste from the consul file. I didn't mean to have that change here, just ended up with this diff after writing the new backend_state file.

**Comment by sslotnick-isp** [comment]
@jbardin Thanks for implementing this. I may have found an issue unless I am misunderstanding the way environments are supposed to work. Are environments supposed to be shared across different services? That is, if I have 2 separate sets of terraform configuration but both point to the same bucket, should I see the same environments across both? Because this change puts `env:` at the root of the bucket, any other service configured to use that bucket will see the same environments. I'm not sure this was intentional because it looks to differ from the Consul equivalent. If this was unintentional I am happy to open a proper issue for it (or submit a patch).

This roughly demonstrates the issue:
```
$ cd serviceA/
$ terraform init -backend=true -backend-config="bucket=my-bucket" -backend-config="region=us-west-2" -backend-config="key=terraform/serviceA"`
$ terraform env list
 * default
$ terraform env new NewEnv

$ cd ../serviceB
$ terraform init -backend=true -backend-config="bucket=my-bucket" -backend-config="region=us-west-2" -backend-config="key=terraform/serviceB"`
$ terraform env list
 * default
    NewEnv
```

The S3 hierarchy ends up looking like this:
```
my-bucket/terraform/serviceA
my-bucket/terraform/serviceB
my-bucket/env:/NewEnv/terraform/serviceA
my-bucket/env:/NewEnv/terraform/serviceB
```

**Comment by jbardin** [comment]
@sslotnick-isp,

The decision was intentional, putting the environments at the same configuration level as the bucket rather than the path. I don't see a hard reason to keep this though, and would be good to match the behavior of the 2 environment-enabled backends.

I'll put up a PR ASAP.

**Comment by sslotnick-isp** [comment]
@jbardin, Was a fix ever created for this? Let me know if I should create an issue or submit a PR.

**Comment by jbardin** [comment]
@sslotnick-isp,

Sorry, there was (#13104) but it's on hold for the moment. 
Because there are states in the wild using the current setup, there needs to be a migration plan, and a safe migration has a few edge cases here that need to be covered. I hope to be able to finish that at some point during the next release cycle, but it's not as simple as just renaming the states.

**Comment by sslotnick-isp** [comment]
@jbardin Thanks -- I kind of figured as much. One option might be adding a backend config flag for "shared environments" that determines if the env key is a prefix or suffix. While it doesn't match Consul or work for my use case, the prefix naming convention might be interesting for some users in the future.

**Pull request merged**